sbi/inference/abc/smcabc.py

```diff
@@ -0,0 +1,487 @@
+from __future__ import annotations
+
+from pyro.distributions.empirical import Empirical
+from sbi.inference.abc.abc_base import ABCBASE
+from typing import Callable, Optional, Tuple, Union
+
+import torch
+from torch import Tensor, ones, tensor
+from torch.distributions import Distribution, Multinomial, MultivariateNormal
+from pyro.distributions import Uniform
+from numpy import ndarray
+import logging
+
+
+class SMCABC(ABCBASE):
+    def __init__(
+        self,
+        simulator: Callable,
+        prior: Distribution,
+        x_o: Union[Tensor, ndarray],
+        distance: Union[str, Callable] = "l2",
+        num_workers: int = 1,
+        simulation_batch_size: int = 1,
+        show_progress_bars: bool = True,
+        kernel: Optional[str] = "gaussian",
+        algorithm_variant: str = "C",
+    ):
+        r"""Sequential Monte Carlo Approximate Bayesian Computation.
+
+        We distinguish between three different SMC methods here:
+            - A: Toni et al. 2010 (Phd Thesis)
+            - B: Sisson et al. 2007 (with correction from 2009)
+            - C: Beaumont et al. 2009
+
+        In Toni et al. 2010 we find an overview of the differences on page 34:
+            - B: same as A except for resampling of weights if the effective sampling
+                size is too small.
+            - C: same as A except for calculation of the covariance of the perturbation
+                kernel: the kernel covariance is a scaled version of the covariance of
+                the previous population.
+
+        Args:
+            kernel: Perturbation kernel.
+            algorithm_variant: Indicating the choice of algorithm variant, A, B, or C.
+
+        See docstring of `PosteriorEstimator` class for all other arguments.
+        """
+
+        super().__init__(
+            simulator=simulator,
+            prior=prior,
+            x_o=x_o,
+            distance=distance,
+            num_workers=num_workers,
+            simulation_batch_size=simulation_batch_size,
+            show_progress_bars=show_progress_bars,
+        )
+
+        kernels = ("gaussian", "uniform")
+        assert (
+            kernel in kernels
+        ), f"Kernel '{kernel}' not supported. Choose one from {kernels}."
+        self.kernel = kernel
+
+        algorithm_variants = ("A", "B", "C")
+        assert algorithm_variant in algorithm_variants, (
+            f"SMCABC variant '{algorithm_variant}' not supported, choose one from"
+            " {algorithm_variants}."
+        )
+        self.algorithm_variant = algorithm_variant
+        self.distance_to_x0 = lambda x: self.distance(self.x_o, x)
+        self.num_simulations = 0
+        self.num_simulation_budget = 0
+
+        self.logger = logging.getLogger(__name__)
+
+        # Define simulator that keeps track of budget.
+        def simulate_with_budget(theta):
+            self.num_simulations += theta.shape[0]
+            return self._batched_simulator(theta)
```

Comment:
do we need this or is this a benchmark leftover?

Reply:
We need this. otherwise SMC will just keep simulating until the enough particles are accepted. if eps is very small this can result in millions of simulations.